src/core/qfieldcloudprojectsmodel.cpp

```diff
@@ -359,262 +403,596 @@ void QFieldCloudProjectsModel::cancelDownloadProject( const QString &projectId )
+      return;
+    }
+
+    const QJsonObject payload = QJsonDocument::fromJson( rawReply->readAll() ).object();
+    const QString packageId = payload.value( QStringLiteral( "package_id" ) ).toString();
+    const QString packagedAt = payload.value( QStringLiteral( "packaged_at" ) ).toString();
+
+    if (
+      packageId.isNull()
+      || packagedAt.isNull()
+      || !payload.value( QStringLiteral( "files" ) ).isArray() )
+    {
+      QgsLogger::debug( QStringLiteral( "Project %1: JSON for package does not contain the expected fields: package_id(string), packaged_at(string), files(array), layers(object)" ).arg( projectId ) );
+      emit projectDownloadFinished( projectId, tr( "Latest package data response error." ) );
+      return;
     }
+
+    const QJsonArray files = payload.value( QStringLiteral( "files" ) ).toArray();
+    for ( const QJsonValue &fileValue : files )
+    {
+      QJsonObject fileObject = fileValue.toObject();
+      int fileSize = fileObject.value( QStringLiteral( "size" ) ).toInt();
+      QString fileName = fileObject.value( QStringLiteral( "name" ) ).toString();
+      QString projectFileName = QStringLiteral( "%1/%2/%3/%4" ).arg( QFieldCloudUtils::localCloudDirectory(), mUsername, projectId, fileName );
+      QString cloudChecksum = fileObject.value( QStringLiteral( "sha256" ) ).toString();
+      QString localChecksum = FileUtils::fileChecksum( projectFileName, QCryptographicHash::Sha256 ).toHex();
+
+      if (
+        !fileObject.value( QStringLiteral( "size" ) ).isDouble()
+        || fileName.isEmpty()
+        || cloudChecksum.isEmpty() )
+      {
+        QgsLogger::debug( QStringLiteral( "Project %1: package in \"files\" list does not contain the expected fields: size(int), name(string), sha256(string)" ).arg( projectId ) );
+        emit projectDownloadFinished( projectId, tr( "Latest package data structure error." ) );
+        return;
+      }
+
+      if ( cloudChecksum == localChecksum )
+        continue;
+
+      project->downloadFileTransfers.insert( fileName, FileTransfer( fileName, fileSize ) );
+      project->downloadBytesTotal += std::max( fileSize, 0 );
+    }
+
+    const QJsonObject layers = payload.value( QStringLiteral( "layers" ) ).toObject();
+    bool hasLayerExportErrror = false;
+    for ( const QString &layerKey : layers.keys() )
+    {
+      QJsonObject layer = layers.value( layerKey ).toObject();
+      QString layerName = layer.value( QStringLiteral( "name" ) ).toString();
+      QString layerStatus = layer.value( QStringLiteral( "error_code" ) ).toString();
+
+      if (
+        layerKey.isEmpty()
+        || layerName.isEmpty()
+        || layerStatus.isEmpty()
+        || !layer.value( QStringLiteral( "is_valid" ) ).isBool() )
+      {
+        QgsLogger::debug( QStringLiteral( "Project %1: JSON structure package in \"files\" list does not contain the expected fields: name(string), status(string), valid(bool)" ).arg( projectId ) );
```

Comment:
Shouldn't this refer to "layers"?

Reply:
@nirvn do you mean `layer.value( QStringLiteral( "is_valid" ) ).isBool()` ? It should be an individual `layer` object, as `layers` is an array.

Reply:
There was another trouble with the debug message, fixed.

Reply:
The whole message was a bit odd, fixed.

Reply:
@suricactus , I meant the debug message seems to be a copy/paste from the section above that iterates through "files", and I was wondering whether the text needed tweaking to refer to "layers".

Reply:
Yes, I realized this. You are way too fast to react. :) Thanks.